radix-engine-interface/src/time/date_time.rs

```diff
@@ -0,0 +1,512 @@
+            remaining_secs += SECONDS_IN_A_DAY as i64;
+            days_since_march_y2k -= 1;
+        }
+
+        let mut num_400_year_cycles = days_since_march_y2k / DAYS_PER_400Y;
+        let mut remaining_days = days_since_march_y2k % DAYS_PER_400Y;
+        if remaining_days < 0 {
+            remaining_days += DAYS_PER_400Y;
+            num_400_year_cycles -= 1;
+        }
+
+        let mut num_100_year_cycles = remaining_days / DAYS_PER_100Y;
+        if num_100_year_cycles == 4 {
+            // Subtract one due 400 years cycle (400 years cycle fits 4 100y cycles)
+            num_100_year_cycles -= 1;
+        }
+        remaining_days -= num_100_year_cycles * DAYS_PER_100Y;
+
+        let mut num_4_year_cycles = remaining_days / DAYS_PER_4Y;
+        if num_4_year_cycles == 25 {
+            // Subtract one due 100 years cycle (100 years cycle fits 25 4y cycles)
+            num_4_year_cycles -= 1;
+        }
+        remaining_days -= num_4_year_cycles * DAYS_PER_4Y;
+
+        let mut remaining_years = remaining_days / 365;
+        if remaining_years == 4 {
+            // Subtract one due to four years cycle
+            remaining_years -= 1;
+        }
+        remaining_days -= remaining_years * 365;
+
+        let mut year =
+            remaining_years +
+                4 * num_4_year_cycles +
+                100 * num_100_year_cycles +
+                400 * num_400_year_cycles
+                + 2000 /* Add the base year (after shifting) */;
+
+        let mut days_in_months_starting_on_march = LEAP_YEAR_DAYS_IN_MONTHS.clone();
+        days_in_months_starting_on_march.rotate_left(2);
+
+        let mut month = 0;
+        while days_in_months_starting_on_march[month] as i64 <= remaining_days {
+            remaining_days -= days_in_months_starting_on_march[month] as i64;
+            month += 1;
+        }
+
+        // Shift the month back to Jan
+        // Handle any overflows, in case we need to add another year after shifting
+        month += 2;
+        if month >= 12 {
+            month -= 12;
+            year += 1;
+        }
+
+        // Shift 0-based month to 1-based
+        month += 1;
+
+        // Shift 0-based day of month to 1-based
+        let day_of_month = remaining_days + 1;
+
+        let hour = remaining_secs / SECONDS_IN_AN_HOUR as i64;
+        let minute = remaining_secs / SECONDS_IN_A_MINUTE as i64 % SECONDS_IN_A_MINUTE as i64;
+        let second = remaining_secs % SECONDS_IN_A_MINUTE as i64;
+
+        Ok(Self {
+            year: u32::try_from(year).map_err(|_| "year overflow")?,
+            month: u8::try_from(month).map_err(|_| "month overflow")?,
+            day_of_month: u8::try_from(day_of_month).map_err(|_| "day_of_month overflow")?,
+            hour: u8::try_from(hour).map_err(|_| "hour overflow")?,
+            minute: u8::try_from(minute).map_err(|_| "minute overflow")?,
+            second: u8::try_from(second).map_err(|_| "second overflow")?,
```

Comment:
Shouldn't no-error be expected here?

Reply:
true, none of these should ever overflow

Reply:
(changed to `expect`)